finality-aleph/src/runtime_api.rs

```diff
@@ -63,36 +64,74 @@ where
         }
     }
 
-    fn read_storage<D: Decode>(
+    fn access_storage<D: Decode>(
+        &self,
+        storage_key: Vec<u8>,
+        at_block: BlockHash,
+    ) -> Result<D, ApiError> {
+        let encoded = self
+            .client
+            .storage(at_block, &sc_client_api::StorageKey(storage_key))
+            .map_err(|_| ApiError::StorageAccessFailure)?
+            .ok_or(ApiError::NoStorage)?;
+        D::decode_all(&mut encoded.0.as_ref()).map_err(ApiError::DecodeError)
+    }
+
+    fn read_storage_value<D: Decode>(
         &self,
         pallet: &str,
         item: &str,
         at_block: BlockHash,
     ) -> Result<D, ApiError> {
         let storage_key = [twox_128(pallet.as_bytes()), twox_128(item.as_bytes())].concat();
+        match self.access_storage(storage_key, at_block) {
+            Err(ApiError::NoStorage) => Err(ApiError::NoStorageValue(pallet.into(), item.into())),
+            other => other,
+        }
+    }
 
-        let encoded = match self
-            .client
-            .storage(at_block, &sc_client_api::StorageKey(storage_key))
-        {
-            Ok(Some(e)) => e,
-            _ => return Err(ApiError::NoStorage(pallet.to_string(), item.to_string())),
-        };
-
-        D::decode_all(&mut encoded.0.as_ref()).map_err(ApiError::DecodeError)
+    #[allow(dead_code)]
+    fn read_storage_map<H: StorageHasher, D: Decode, E: Encode>(
+        &self,
+        pallet: &str,
+        item: &str,
+        key: E,
+        at_block: BlockHash,
+    ) -> Result<D, ApiError> {
+        let mut storage_key = [twox_128(pallet.as_bytes()), twox_128(item.as_bytes())].concat();
+        let p = key.using_encoded(H::hash);
+        storage_key.extend(p.as_ref());
+        match self.access_storage::<D>(storage_key, at_block) {
+            Err(ApiError::NoStorage) => {
+                Err(ApiError::NoStorageMapElement(pallet.into(), item.into()))
+            }
+            other => other,
+        }
     }
 }
 
 #[derive(Clone, Debug)]
 pub enum ApiError {
-    NoStorage(String, String),
+    StorageAccessFailure,
+    NoStorage,
```

Comment:
Yeah, so in `access_storage` we don't have the `String` to pass to the `NoStorage` error, we only have the byte array which shouldn't be logged in `Display`. Thus I have introduced `NoStorage` only to be propagated from `access_storage` to `read_storage_map` and `read_storage_value`, so that they can convert it into a more specific `NoStorage...` error which can be given the `String` which represents the "unencoded key" to the storage. It's not the most beautiful way, I know, but the other options I considered (returning `Result<Option, ApiError>` from `access_storage` and converting `None` to `NoStorage` on the upper level, or introducing a new error type for `access_storage`) seemed like just too complicated.

Reply:
Ok, we can leave it then